tests/evm/test_calldatacopy.py

```diff
@@ -0,0 +1,311 @@
+    src_addr: int,
+    dst_addr: int,
+    src_addr_end: int,
+    bytes_left: int,
+    from_tx: bool,
+    rw_counter: int,
+    program_counter: int,
+    stack_pointer: int,
+    memory_size: int,
+    gas_left: int,
+    code_source: RLC,
+) -> Tuple[StepState, Sequence[RW]]:
+    aux_data = CopyToMemoryAuxData(
+        src_addr=src_addr,
+        dst_addr=dst_addr,
+        src_addr_end=src_addr_end,
+        bytes_left=bytes_left,
+        from_tx=from_tx,
+        tx_id=TX_ID,
+    )
+    step = StepState(
+        execution_state=ExecutionState.CopyToMemory,
+        rw_counter=rw_counter,
+        call_id=1,
+        is_root=from_tx,
+        program_counter=program_counter,
+        stack_pointer=stack_pointer,
+        gas_left=gas_left,
+        memory_size=memory_size,
+        code_source=code_source,
+        aux_data=aux_data,
+    )
+
+    rws = []
+    num_bytes = min(MAX_COPY_BYTES, bytes_left)
+    for i in range(num_bytes):
+        byte = buffer_map[src_addr + i] if src_addr + i < src_addr_end else 0
+        if not from_tx and src_addr + i < src_addr_end:
+            rws.append(
+                (
+                    rw_counter,
+                    RW.Read,
+                    RWTableTag.Memory,
+                    CALL_ID,
+                    src_addr + i,
+                    0,
+                    byte,
+                    0,
+                    0,
+                    0,
+                )
+            )
+            rw_counter += 1
+        rws.append(
+            (
+                rw_counter,
+                RW.Write,
+                RWTableTag.Memory,
+                CALL_ID,
+                dst_addr + i,
+                0,
+                byte,
+                0,
+                0,
+                0,
+            )
+        )
+        rw_counter += 1
+    return step, rws
+
+
+def make_copy_steps(
+    buffer: bytes,
+    buffer_addr: int,
+    src_addr: int,
+    dst_addr: int,
+    length: int,
+    from_tx: bool,
+    rw_counter: int,
+    program_counter: int,
+    stack_pointer: int,
+    memory_size: int,
+    gas_left: int,
+    code_source: RLC,
+) -> Tuple[Sequence[StepState], Sequence[RW]]:
+    buffer_addr_end = buffer_addr + len(buffer)
+    buffer_map = dict(zip(range(buffer_addr, buffer_addr_end), buffer))
+    steps = []
+    rws = []
+    bytes_left = length
+    while bytes_left > 0:
+        curr_rw_counter = rws[-1][0] + 1 if rws else rw_counter
+        new_step, new_rws = make_copy_step(
+            buffer_map,
+            src_addr,
+            dst_addr,
+            buffer_addr_end,
+            bytes_left,
+            from_tx,
+            curr_rw_counter,
+            program_counter,
+            stack_pointer,
+            memory_size,
+            gas_left,
+            code_source,
+        )
+        steps.append(new_step)
+        rws.extend(new_rws)
+        src_addr += MAX_COPY_BYTES
+        dst_addr += MAX_COPY_BYTES
+        bytes_left -= MAX_COPY_BYTES
```

Comment:
Does make_copy_step guarantee consuming `MAX_COPY_BYTES` all the time? It seems not the case in the last step

Reply:
In the last step, it actually consumes `min(bytes_left, MAX_COPY_BYTES)`. But anyway, it will lead to `bytes_left <= 0` and the [while loop](https://github.com/appliedzkp/zkevm-specs/blob/master/tests/evm/test_calldatacopy.py#L144) will break out.